api/docs/mapstruct.md

```diff
@@ -0,0 +1,608 @@
+
+    final EmployeeModel employeeModel = employeeMapper.toModel(employeeDbEntity);
+    assertThat(employeeModel.getDepartment()).isEqualTo(Department.IT);
+    assertThat(employeeModel.getSalary()).isWithin(1.0e-6).of(6000.0);
+    assertThat(employeeModel.getName()).isEqualTo("John Doe");
+    assertThat(employeeModel.getAddress()).isEqualTo(ADDRESS);
+  }
+}
+```
+## Tips & Strategies
+
+### Stub it so it builds, and go one property at a time
+For big target classes, put in some temporary `ignore` statements or other hacks just to get it
+to the point that it generates the implemmentation class and actually compiles without errors. Then
+proceed one property at a time until all the warnings are gone.
+ 
+### Proofread the generated code
+First, you should *always* look at the generated code and step through it in the debugger. This is
+because, while the codegen should be deterministic, most of us don't go around with its order of
+operator precedence in our head. Particularly if you have many input parameters with multiple fields
+with common names, you may find that `foo.id` is getting used instead of `bar.id` becase one's a 
+`long` and the other an `int`. Or something. 
+
+In general, the library gives very reasonable error
+messages and is very cranky at compile time. This is A Good Thing.
+
+## Write tests for mappers
+It's pretty tedious to write tests for generated code, as the computer is coding faster than you.
+I still think it's worth it, because,
+- it documents usage of your generated methods
+- you catch things the compiler can't (such as when a string format changes)
+- it's a good opportunity to update old test code
+- It makes it look less magical
+
+### Consider other use cases
+It's possible to write a mapper from a type to itself and use that as a copy constructor of sorts.
+
+### Tune the classes' public APIs
+If you find yourself writing arcane helper methods, ask yourself if adding a simple setter or
+getter on one of your classes would make the job easier and the generated code cleaner. Since our
+model classes are generated, we frequently have to have external helper methods defined in either
+the mapper class or a utility class.
+
+## Handling missing target properties
+### Avoid `ignore = true` unless absolutely necessary
+It's annoying to see warnings about missing target properties, and for good reason. The quick way to
+stifle these is to use the `ignore` parameter, which tells MapStruct to elide the setter code for
+that target property.
+```java
+@Mapping(target = "billingAccountName", ignore = true)
+```
+
+While there are some places this is necessary, it should really be a last resort, for a couple of
+reasons.
+
+First, we plan to use mappers to replace existing chunks of code in a clean way. We don't want to
+have to docment that mapping function X leaves the created object in state Y with respect to fields
```

Comment:
docment -> document. does your mac hate u?

Reply:
yeah. My cat was sleeping on the keyboard and I wiped all the hair and whatever oil she likes to secrete off, and the u key has never been the same since.